scrapy_poet/utils.py

```diff
@@ -125,40 +123,56 @@ def is_response_going_to_be_used(request, spider):
     return False
 
 
-def build_plan(callback, instances) -> Tuple[andi.Plan, Dict[Type, PageObjectInputProvider]]:
+def build_plan(callback) -> andi.Plan:
     """Build a plan for the injection in the callback."""
-    provider_instances = build_providers(instances)
-    plan = andi.plan(
+    return andi.plan(
         callback,
         is_injectable=is_injectable,
-        externally_provided=provider_instances.keys()
+        externally_provided=set.union(*(p.provided_classes for p in providers))
     )
-    return plan, provider_instances
 
 
-def build_providers(instances) -> Dict[Type, PageObjectInputProvider]:
-    result = {}
-    for cls, provider in providers.items():
+def build_providers(instances) -> Set[PageObjectInputProvider]:
+    result = set()
+    for provider in providers:
         kwargs = andi.plan(
             provider,
             is_injectable=is_injectable,
             externally_provided=instances.keys(),
             full_final_kwargs=True,
         ).final_kwargs(instances)
-        result[cls] = provider(**kwargs)  # type: ignore
+        result.add(provider(**kwargs))  # type: ignore
 
     return result
 
 
 @inlineCallbacks
-def build_instances(plan: andi.Plan, providers):
+def build_instances(plan: andi.Plan, provider_instances: Set[PageObjectInputProvider]):
     """Build the instances dict from a plan."""
     instances = {}
-    for cls, kwargs_spec in plan:
-        if cls in providers:
-            instances[cls] = yield maybeDeferred_coro(providers[cls])
-        else:
+
+    dependencies = dict(plan.dependencies)
+
+    for provider in provider_instances:
+        # Discover classes being provided by this provider
+        provided_classes = set()
+        for cls in dependencies.keys():
+            if cls in provider.provided_classes:
+                provided_classes.add(cls)
+
+        if provided_classes:
+            # Create a single instance of this provider and associate it
+            # with all classes being currently provided by it
+            instance = yield maybeDeferred_coro(provider, provided_classes)
+
+        for provided_class in provided_classes:
+            instances[provided_class] = instance[provided_class]
+
+    # Build missing dependencies not addressed by previous step
+    for cls, kwargs_spec in dependencies.items():
```

Comment:
Will it make any difference here in this case? 🤔

Reply:
FYI I'm using the `plan.dependencies` list now because of a refactor to simplify this code.

Reply:
I'm back to dictionaries because of another refactor. It made sense just to use them. 

Regarding your concern with the items order, since Python 3.6 dicts keep the insertion order. 

```python
>>> dict([(1, 1), (2, 2), (3, 3)])
{1: 1, 2: 2, 3: 3}
>>> dict([(2, 2), (1, 1), (3, 3)])
{2: 2, 1: 1, 3: 3}
``` 

This project supports Python 3.6 onwards.

Reply:
But note that it didn't become a Python standard until Python 3.7, so different 3.6 CPython implementations could have not implemented them. So I think better not to rely on that, given that is easy to iterate over the list of tuples in any case.

Reply:
Fixed. Marking as resolved.